jmx_prometheus_javaagent/src/main/java/io/prometheus/jmx/JavaAgent.java

```diff
@@ -1,35 +1,139 @@
+        final Throwable exception;
+
+        FileInfo(long lastChangeTs, String fileName) {
+            this(lastChangeTs, fileName, null);
+        }
+
+        public FileInfo(long lastChangeTs, Throwable exception) {
+            this(lastChangeTs, null, exception);
+        }
+
+        public FileInfo(long lastChangeTs, String fileName, Throwable exception) {
+            this.lastChangeTs = lastChangeTs;
+            this.fileName = fileName;
+            this.exception = exception;
+        }
+
+    }
+
+    public static void premain(String agentArgument, Instrumentation instrumentation) throws Exception {
+        final String[] args = agentArgument.split(":");
+        if (args.length < 2) {
+            System.err.println("Usage: -javaagent:/path/to/JavaAgent.jar=<port>:<yaml configuration file>[:<config watch timeout - default 30s>]");
+            System.exit(1);
+        }
+
+        long timeout = 30l;
+
+        if (args.length > 2) {
+            timeout = Long.parseLong(args[2]);
+        }
+
+        final CountDownLatch jmxDone = new CountDownLatch(1);
+        watchTimer.scheduleAtFixedRate(new JmxConfigurer(jmxDone, new File(args[1])), 0l, TimeUnit.SECONDS.toMillis(timeout));
+        DefaultExports.initialize();
+
+        try {
+            jmxDone.await(10, TimeUnit.SECONDS);
+        } catch (InterruptedException e) {
+            throw new RuntimeException("Failed waiting for JMX registry", e);
+        }
+
+        FileInfo info = lastChange.get();
+        if (info.exception != null) {
+            throw new IllegalStateException("Failed to initialise agent", info.exception);
+        }
+
+        int port = Integer.parseInt(args[0]);
+        server = new Server(port);
+        QueuedThreadPool pool = new QueuedThreadPool();
+        pool.setDaemon(true);
+        server.setThreadPool(pool);
+        ServletContextHandler context = new ServletContextHandler();
+        context.setContextPath("/");
+        server.setHandler(context);
+        context.addServlet(new ServletHolder(new MetricsServlet()), "/metrics");
+        server.start();
+    }
+
+    static class JmxConfigurer extends TimerTask {
+        final CountDownLatch startLatch;
+        final File fileToWatch;
+        AtomicReference<Collector> collectorRef = new AtomicReference<>();
+
+        JmxConfigurer(CountDownLatch startLatch, File fileToWatch) {
+            this.startLatch = startLatch;
+            this.fileToWatch = fileToWatch;
+        }
+
+        @Override
+        public void run() {
+            FileReader reader = null;
+            FileInfo info = null;
+            try {
+                FileInfo lastLoad = lastChange.get();
+                long lastModified = fileToWatch.lastModified();
+                if (lastLoad != null && lastLoad.lastChangeTs == lastModified) {
+                    return;
+                }
+
+                //NB: logic below can result in duplicated metric blip for short time
```

Comment:
As the author and maintainer of client_java, this is not the manner in which the Registry is meant to be used. I'm not aware of anyone using it as you suggest in Java or in any other language or exporter.

The registry allows for concurrent registration so that the user doesn't have to worry about thread safety in the standard use case.

Please change this to handle the swap inside the collector as I've previously requested.